File: registry/examples/registry_ping.py
#!/usr/bin/env python3
"""
Registry Ping - Client that looks up pong via GlobalRegistry

Run GlobalRegistry first (python -m actors.registry), then registry_pong.py,
then run this.

Usage:
    python registry_ping.py [registry_endpoint]

THIS SOFTWARE IS OPEN SOURCE UNDER THE MIT LICENSE

Copyright 2025 Vincent Maciejewski, & M2 Tech
"""

import os
import signal
import sys

# Add actors package to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'python'))

from actors import (
    Actor, Envelope, Manager, Start, Shutdown,
    ZmqSender, ZmqReceiver, RemoteActorRef, register_message,
    RegistryClient, RegistryError, ActorNotFoundError, ActorOfflineError
)


# Define Ping message
@register_message
class Ping:
    def __init__(self, count: int = 0):
        self.count = count


# Define Pong message
@register_message
class Pong:
    def __init__(self, count: int = 0):
        self.count = count


class PingActor(Actor):
    """Sends Ping to pong (looked up via registry), receives Pong back."""

    def __init__(self, manager_handle, registry_client: RegistryClient, zmq_sender: ZmqSender):
        super().__init__()
        self.manager_handle = manager_handle
        self.registry_client = registry_client
        self.zmq_sender = zmq_sender
        self.pong_ref = None

    def on_start(self, env: Envelope) -> None:
        print("PingActor: Starting ping-pong...")

        try:
            # Look up pong actor via registry
            pong_endpoint = self.registry_client.lookup("pong")
            print(f"PingActor: Found 'pong' at {pong_endpoint}")

            # Create remote ref for pong
            self.pong_ref = RemoteActorRef("pong", pong_endpoint, self.zmq_sender)

            # Send first ping
            print("PingActor: Sending first ping")
            self.pong_ref.send(Ping(1), sender=self._actor_ref)

        except ActorNotFoundError as e:
            print(f"PingActor: Failed to find 'pong': {e}")
            print("Make sure registry_pong.py is running first!")
            self.manager_handle.terminate()
        except ActorOfflineError as e:
            print(f"PingActor: 'pong' is offline: {e}")
            self.manager_handle.terminate()
        except RegistryError as e:
            print(f"PingActor: Registry error: {e}")
            self.manager_handle.terminate()

    def on_pong(self, env: Envelope) -> None:
        print(f"PingActor: Received pong {env.msg.count} from remote")

        if env.msg.count >= 5:
            print("PingActor: Done!")
            self.manager_handle.terminate()
        else:
            # Send next ping
            if self.pong_ref:
                self.pong_ref.send(Ping(env.msg.count + 1), sender=self._actor_ref)


def main():
    registry_endpoint = "tcp://localhost:5555"
    if len(sys.argv) > 1:
        registry_endpoint = sys.argv[1]

    ENDPOINT = "tcp://*:5002"
    LOCAL_ENDPOINT = "tcp://localhost:5002"

    print("=== Registry Ping Process (port 5002) ===")
    print(f"Registry: {registry_endpoint}")

    # Create manager
    mgr = Manager(endpoint=LOCAL_ENDPOINT)
    handle = mgr.get_handle()

    # Create ZMQ sender/receiver for remote communication
    zmq_sender = ZmqSender(local_endpoint=LOCAL_ENDPOINT)
    zmq_receiver = ZmqReceiver(ENDPOINT, mgr, zmq_sender)

    # Create registry client
    registry_client = RegistryClient("PingManager", registry_endpoint)

    try:
        registry_client.start_heartbeat()
    except RegistryError as e:
        print(f"Failed to connect to registry: {e}")
        return 1

    # Create ping actor - it will lookup pong via registry in on_start
    ping_actor = PingActor(handle, registry_client, zmq_sender)

    # Register actors with manager
    mgr.manage("zmq_receiver", zmq_receiver)
    mgr.manage("ping", ping_actor)

    # Set up signal handler for clean shutdown
    def signal_handler(sig, frame):
        print("\nShutting down...")
        handle.terminate()

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    # Start actors
    mgr.init()
    print("Ping process starting...")

    try:
        mgr.run()
    except KeyboardInterrupt:
        pass

    # Cleanup
    registry_client.close()
    mgr.end()

    print("=== Registry Ping Process Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: registry/examples/registry_ping.rs
/*
Registry Ping - Client that looks up pong via GlobalRegistry

Run GlobalRegistry first (python -m actors.registry), then registry_pong,
then run this.

Usage:
    cd ~/actors/rust
    cargo run --example registry_ping

THIS SOFTWARE IS OPEN SOURCE UNDER THE MIT LICENSE

Copyright 2025 Vincent Maciejewski, & M2 Tech
*/

use std::sync::Arc;

use serde::{Deserialize, Serialize};

use actors::{
    Actor, ActorContext, ActorRef, Manager, ThreadConfig, ManagerHandle,
    ZmqReceiver, ZmqSender, RemoteActorRef, register_remote_message, define_message, handle_messages,
    Start,
    RegistryClient, RegistryError,
};

// Define messages
#[derive(Serialize, Deserialize, Default)]
struct Ping {
    count: i32,
}
define_message!(Ping);

#[derive(Serialize, Deserialize, Default)]
struct Pong {
    count: i32,
}
define_message!(Pong);

/// PingActor - Sends Ping to pong (looked up via registry), receives Pong back
struct PingActor {
    manager_handle: ManagerHandle,
    registry_client: Arc<RegistryClient>,
    zmq_sender: Arc<ZmqSender>,
    pong_ref: Option<ActorRef>,
}

impl PingActor {
    fn new(
        manager_handle: ManagerHandle,
        registry_client: Arc<RegistryClient>,
        zmq_sender: Arc<ZmqSender>,
    ) -> Self {
        PingActor {
            manager_handle,
            registry_client,
            zmq_sender,
            pong_ref: None,
        }
    }
}

handle_messages!(PingActor,
    Start => on_start,
    Pong => on_pong
);

impl PingActor {
    fn on_start(&mut self, _msg: &Start, ctx: &mut ActorContext) {
        println!("PingActor: Starting ping-pong...");

        // Look up pong actor via registry
        match self.registry_client.lookup("pong") {
            Ok(endpoint) => {
                println!("PingActor: Found 'pong' at {}", endpoint);

                // Create remote ref for pong
                let pong_ref = RemoteActorRef::new("pong", &endpoint, Arc::clone(&self.zmq_sender));
                self.pong_ref = Some(pong_ref.into_actor_ref());

                // Send first ping
                println!("PingActor: Sending first ping");
                if let Some(ref pong) = self.pong_ref {
                    pong.send(Box::new(Ping { count: 1 }), ctx.self_ref());
                }
            }
            Err(RegistryError::NotFound(_)) => {
                eprintln!("PingActor: Failed to find 'pong'");
                eprintln!("Make sure registry_pong is running first!");
                self.manager_handle.terminate();
            }
            Err(RegistryError::Offline(_)) => {
                eprintln!("PingActor: 'pong' is offline");
                self.manager_handle.terminate();
            }
            Err(e) => {
                eprintln!("PingActor: Registry error: {}", e);
                self.manager_handle.terminate();
            }
        }
    }

    fn on_pong(&mut self, msg: &Pong, ctx: &mut ActorContext) {
        println!("PingActor: Received pong {} from remote", msg.count);

        if msg.count >= 5 {
            println!("PingActor: Done!");
            self.manager_handle.terminate();
        } else {
            // Send next ping
            if let Some(ref pong) = self.pong_ref {
                pong.send(Box::new(Ping { count: msg.count + 1 }), ctx.self_ref());
            }
        }
    }
}

fn main() {
    let registry_endpoint = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "tcp://localhost:5555".to_string());

    let endpoint = "tcp://0.0.0.0:5002";
    let local_endpoint = "tcp://localhost:5002";

    println!("=== Registry Ping Process (port 5002) ===");
    println!("Registry: {}", registry_endpoint);

    // Register messages for remote serialization
    register_remote_message::<Ping>("Ping");
    register_remote_message::<Pong>("Pong");

    // Create registry client
    let registry_client = Arc::new(RegistryClient::new("PingManager", &registry_endpoint));
    registry_client.start_heartbeat();

    // Create ZMQ sender for remote communication
    let zmq_sender = Arc::new(ZmqSender::new(local_endpoint));

    // Create manager
    let mut mgr = Manager::new();
    let handle = mgr.get_handle();

    // Create ping actor - it will lookup pong via registry in on_start
    let ping_actor = PingActor::new(
        handle.clone(),
        Arc::clone(&registry_client),
        Arc::clone(&zmq_sender),
    );
    let ping_ref = mgr.manage("ping", Box::new(ping_actor), ThreadConfig::default());

    // Create ZMQ receiver and register local actors
    let zmq_receiver = ZmqReceiver::new(endpoint, Arc::clone(&zmq_sender));
    zmq_receiver.register("ping", ping_ref);

    // Start the receiver
    let mut receiver_handle = zmq_receiver.start();

    // Set up signal handler
    let handle_clone = handle.clone();
    ctrlc::set_handler(move || {
        println!("\nShutting down...");
        handle_clone.terminate();
    }).expect("Error setting Ctrl-C handler");

    mgr.init();
    println!("Ping process starting...");

    // Wait until terminated
    mgr.run();

    // Cleanup
    receiver_handle.stop();
    mgr.end();

    println!("=== Registry Ping Process Complete ===");
}
